[FLINK-11068][table] Convert the API classes of *Table to interfaces

hequn8128 opened the pull request:
## What is the purpose of the change

This pull request convert table related classes into interface.


## Brief change log

  - Convert Table, GroupedTable, GroupWindowedTable, OverWindowedTable, WindowGroupedTable into interface.
  - Port TemporalTableFunction into flink-common. I add no methods in TemporalTableFunction. It seems we don't need to add methods in `TemporalTableFunction` for now. We can add it later if we find it necessary. It's easier and safer for adding methods than removing methods.


## Verifying this change

This change is already covered by existing tests, such as tests for tableApi. And changes for temporal table has been covered by TemporalTableJoin UT/IT test cases.


## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)

twalthr commented:
@flinkbot approve-until consensus

dawidwys commented on flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/GroupWindowedTable.java:
```diff
@@ -0,0 +1,69 @@
+	 * <p>Aggregations are performed per group and defined by a subsequent {@code select(...)}
+	 * clause similar to SQL SELECT-GROUP-BY query.
+	 *
+	 * <p>Example:
```
How about we change it to `Scala example:`? This will not work in Java.

dawidwys commented on flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/GroupedTable.java:
```diff
@@ -0,0 +1,57 @@
+	 * Performs a selection operation on a grouped table. Similar to an SQL SELECT statement.
+	 * The field expressions can contain complex expressions and aggregations.
+	 *
+	 * <p>Example:
```
Scala example:?

dawidwys commented on flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/OverWindowedTable.java:
```diff
@@ -0,0 +1,61 @@
+	 * Performs a selection operation on a over windowed table. Similar to an SQL SELECT statement.
+	 * The field expressions can contain complex expressions and aggregations.
+	 *
+	 * <p>Example:
```
Scala example:?

dawidwys commented on flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/Table.java:
```diff
@@ -0,0 +1,922 @@
+ * <p>Operations such as {@code join}, {@code select}, {@code where} and {@code groupBy} either
+ * take arguments in a Scala DSL or as an expression String. Please refer to the documentation for
+ * the expression syntax.
+ *
```
nit: empty line

dawidwys commented on flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/Table.java:
```diff
@@ -0,0 +1,922 @@
+	 * Performs a selection operation. Similar to an SQL SELECT statement. The field expressions
+	 * can contain complex expressions and aggregations.
+	 *
+	 * <p>Example:
```
Scala example:?

dawidwys commented on flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/Table.java:
```diff
@@ -0,0 +1,922 @@
+	 * Renames the fields of the expression result. Use this to disambiguate fields before
+	 * joining to operations.
+	 *
+	 * <p>Example:
```
Scala example:?

dawidwys commented on flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/Table.java:
```diff
@@ -0,0 +1,922 @@
+	 * Filters out elements that don't pass the filter predicate. Similar to a SQL WHERE
+	 * clause.
+	 *
+	 * <p>Example:
```
Scala example:?

dawidwys commented on flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/Table.java:
```diff
@@ -0,0 +1,922 @@
+	 * Filters out elements that don't pass the filter predicate. Similar to a SQL WHERE
+	 * clause.
+	 *
+	 * <p>Example:
```
Scala example:?

dawidwys commented on flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/Table.java:
```diff
@@ -0,0 +1,922 @@
+	 * Groups the elements on some grouping keys. Use this before a selection with aggregations
+	 * to perform the aggregation on a per-group basis. Similar to a SQL GROUP BY statement.
+	 *
+	 * <p>Example:
```
Scala example:?

dawidwys commented on flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/Table.java:
```diff
@@ -0,0 +1,922 @@
+	 *
+	 * <p>Note: Both tables must be bound to the same {@code TableEnvironment} .
+	 *
+	 * <p>Example:
```
How about we use the Java's api by default (the String approach for now)?

dawidwys commented on flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/Table.java:
```diff
@@ -0,0 +1,922 @@
+	 *
+	 * <p>Note: Both tables must be bound to the same {@code TableEnvironment} .
+	 *
+	 * <p>Example:
```
Scala example:?

dawidwys commented on flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/Table.java:
```diff
@@ -0,0 +1,922 @@
+	 *
+	 * <pre>
+	 * {@code
+	 *   left.leftOuterJoin(right).select('a, 'b, 'd)
```
String based condition by default?

dawidwys commented on flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/Table.java:
```diff
@@ -0,0 +1,922 @@
+	 * <p>Note: Both tables must be bound to the same {@code TableEnvironment} and its
+	 * {@code TableConfig} must have null check enabled (default).
+	 *
+	 * <p>Example:
```
Scala example?

dawidwys commented on flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/Table.java:
```diff
@@ -0,0 +1,922 @@
+	 *
+	 * <pre>
+	 * {@code
+	 *   left.rightOuterJoin(right, "a = b").select('a, 'b, 'd)
```
Let's try to stay in the same API approach.

dawidwys commented on flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/Table.java:
```diff
@@ -0,0 +1,922 @@
+	 *
+	 * <pre>
+	 * {@code
+	 *   left.fullOuterJoin(right, "a = b").select('a, 'b, 'd)
```
Let's try to stay in the same API approach.

dawidwys commented on flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/Table.java:
```diff
@@ -0,0 +1,922 @@
+	 *             {@link Table#insertInto(String tableName)}.
+	 */
+	@Deprecated
+	<T> void writeToSink(TableSink<T> sink);
```
Maybe let's just drop it since it's deprecated since 1.7?

dawidwys commented on flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/Table.java:
```diff
@@ -0,0 +1,922 @@
+	 *             {@link Table#insertInto(String tableName, QueryConfig conf)}.
+	 */
+	@Deprecated
+	<T> void writeToSink(TableSink<T> sink, QueryConfig conf);
```
Let's drop it.

dawidwys commented on flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/Table.java:
```diff
@@ -0,0 +1,922 @@
+	 * <p>For batch tables of finite size, windowing essentially provides shortcuts for time-based
+	 * groupBy.
+	 *
+	 * <p>__Note__: Computing windowed aggregates on a streaming table is only a parallel operation
```
```suggestion
	 * <p><b>Note</b>: Computing windowed aggregates on a streaming table is only a parallel operation
```

dawidwys commented on flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/Table.java:
```diff
@@ -0,0 +1,922 @@
+	 * }
+	 * </pre>
+	 *
+	 * <p>__Note__: Computing over window aggregates on a streaming table is only a parallel
```
```suggestion
	 * <p><b>Note</b>: Computing over window aggregates on a streaming table is only a parallel
```

dawidwys commented on flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/Table.java:
```diff
@@ -0,0 +1,922 @@
+	 * operation if the window is partitioned. Otherwise, the whole stream will be processed by a
+	 * single task, i.e., with parallelism 1.
+	 *
+	 * <p>__Note__: Over-windows for batch tables are currently not supported.
```
```suggestion
	 * <p><b>Note</b>: Over-windows for batch tables are currently not supported.
```

dawidwys commented on flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/Table.java:
```diff
@@ -0,0 +1,922 @@
+	 *
+	 * @param fetch the number of records to return. Fetch must be >= 0.
+	 */
+	Table fetch(Integer fetch);
```
How about: 
```suggestion
	Table fetch(int fetch);
```

dawidwys commented on flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/Table.java:
```diff
@@ -0,0 +1,922 @@
+	 *
+	 * @param offset number of records to skip
+	 */
+	Table offset(Integer offset);
```
How about:
```suggestion
	Table offset(int offset);
```

dawidwys commented:
One more thing could we also already remove the javadocs from `TableImpl`? We shouldn't duplicate it.

dawidwys commented on flink-table/flink-sql-client/src/main/java/org/apache/flink/table/client/gateway/local/ResultStore.java:
```diff
@@ -61,7 +60,7 @@ public ResultStore(Configuration flinkConfig) {
 	public <T> DynamicResult<T> createResult(Environment env, TableSchema schema, ExecutionConfig config) {
 
-		final TypeInformation<Row> outputType = Types.ROW_NAMED(schema.getFieldNames(), schema.getFieldTypes());
+		final RowTypeInfo outputType = (RowTypeInfo) Types.ROW_NAMED(schema.getFieldNames(), schema.getFieldTypes());
```
Maybe let's just replace it with the `RowTypeInfo` ctor, if we expect that type anyway?

dawidwys commented on flink-table/flink-sql-client/src/main/java/org/apache/flink/table/client/gateway/local/result/MaterializedCollectBatchResult.java:
```diff
@@ -53,11 +54,13 @@
 
 		accumulatorName = new AbstractID().toString();
-		tableSink = new CollectBatchTableSink(accumulatorName, outputType.createSerializer(config));
+		tableSink =
```
How about we change the return type of `configure` in `CollectBatchTableSink` to `CollectBatchTableSink`?

dawidwys commented on flink-table/flink-sql-client/src/main/java/org/apache/flink/table/client/gateway/local/result/CollectStreamResult.java:
```diff
@@ -72,7 +73,9 @@ public CollectStreamResult(TypeInformation<Row> outputType, ExecutionConfig conf
 		// pass binding address and port such that sink knows where to send to
-		collectTableSink = new CollectStreamTableSink(iterator.getBindAddress(), iterator.getPort(), serializer);
+		collectTableSink =
+			(CollectStreamTableSink) new CollectStreamTableSink(iterator.getBindAddress(), iterator.getPort(), serializer)
```
Same as in the batch case.

dawidwys commented on flink-table/flink-sql-client/src/test/java/org/apache/flink/table/client/gateway/local/result/MaterializedCollectStreamResultTest.java:
```diff
@@ -42,7 +42,7 @@
 	@Test
 	public void testSnapshot() throws UnknownHostException {
-		final TypeInformation<Row> type = Types.ROW(Types.STRING, Types.LONG);
+		final RowTypeInfo type = (RowTypeInfo) Types.ROW(Types.STRING, Types.LONG);
```
Let's go directly for the ctor.

dawidwys commented on flink-table/flink-sql-client/src/test/java/org/apache/flink/table/client/gateway/local/result/MaterializedCollectStreamResultTest.java:
```diff
@@ -90,7 +90,7 @@ public void testSnapshot() throws UnknownHostException {
 	@Test
 	public void testLimitedSnapshot() throws UnknownHostException {
-		final TypeInformation<Row> type = Types.ROW(Types.STRING, Types.LONG);
+		final RowTypeInfo type = (RowTypeInfo) Types.ROW(Types.STRING, Types.LONG);
```
Let's go directly for the ctor.

hequn8128 commented:
@dawidwys  Thanks a lot for your review and valuable suggestions. I have addressed all your comments and updated the PR. Would be great if you can take another look. Thank you.

dawidwys commented:
Thank you a lot @hequn8128 for the work. It looks good. Last minor thing is we should add `override` keywords to the methods of `TableImpl`, but I can do that during merging. Will do that as soon as travis gives green.

dawidwys commented:
@flinkbot approve all

hequn8128 commented:
@dawidwys I added the override methods and fix one checkstyle problem. Sorry for the trouble that brings to you.
